### ISSUE TITLE
Parse CSV file with products

### ISSUE THREAD

**Issue by olivermrbl** [medusajs/medusa#1486]
**Depends on**
- #1485

Subtask of #1265 

**Goal**
Parse CSV file with products in [default template](https://www.notion.so/medusajs/Default-Product-Import-template-40ceb2271c8d47d1ae1aee7e3ec3debb).

**DoD**
- [x] Given a CSV file of products should return array of products and product variants as described in [discussion](https://github.com/medusajs/medusa/discussions/1272)

**Comments**
When importing, we merge the “product columns” to form the data needed for the product part of the import and we then create multiple variants.

Some fields are required and some are optional; note that the column width is dynamic as there may be one or more product options/images/prices. We should potentially match the product columns by name. When exporting we should dynamically figure out how many columns to include in order to include all the data.

It should be possible to use the above format for updating existing data, matching product data by “Product Handle” (or maybe “External ID”), and product variant data by SKU. If only updating variant data it should be allowed to just upload a file with the columns that should be updated. E.g. SKU, Inventory Quantity.

The required fields for a new product are:

- handle, title
- Other fields that have defaults if not provided:
    - Status - draft
    - Product Option 1 - Default Option
    - Profile - Default Profile
    - Discountable - true
    
The required fields for a new product variant are:

- If product has product option then Variant Option Value must be defined
- Other fields that have defaults if not provided:
    - Title - concat of option values
    - Allow Backorder - false
    - Manage Inventory - true
    - Variant Option Value 1 (if not defined on product) - Default Variant

**Comment by adrien2p** [comment]
Hey guys, I was writing up a draft example on an approach for the csv parsing and import.
Here is what I've done during launch, it is only meant to give an idea of the approach and in any case it can work as it it.
Fill free to give your feedback :)

```ts
import { AwilixContainer } from "awilix"

type CsvParserContext = {
  line: number
  column: string
}

type CsvSchema = {
  columns: {
    name: string
    mapToEntityField: string
    validator: (
      value: string,
      context: CsvParserContext,
      container: AwilixContainer
    ) => Promise<unknown | never>
  }[]
}

interface IParser<TResult> {
  parse(filePath: string): Promise<TResult[]>
}

interface ICsvParser<TResult> extends IParser<TResult> {
  $$delimiter: string
}

abstract class AbstractParser<TSchema, TParserResult, TOutputResult>
  implements IParser<TParserResult>
{
  protected $$schema: TSchema

  protected constructor(schema: TSchema) {
    this.$$schema = schema
  }

  public abstract parse(filePath: string): Promise<TParserResult[]>
  public abstract readData(
    fileContent: TParserResult[]
  ): IterableIterator<TOutputResult>
}

/* ==================================================================================================*/

import someParser from "someParser"
import { readFile } from "fs/promises"

abstract class CsvParser<
    TSchema extends CsvSchema = CsvSchema,
    TParserResult = any,
    TOutputResult = any
  >
  extends AbstractParser<TSchema, TParserResult, TOutputResult>
  implements ICsvParser<TParserResult>
{
  $$delimiter = ";"

  constructor(protected readonly container: AwilixContainer, schema: TSchema) {
    super(schema)
  }

  public async parse(filePath: string): Promise<TParserResult[]> {
    const content = await readFile(filePath)
    return await someParser.parse(content.toString())
  }

  abstract readData(
    parsedContent: TParserResult[]
  ): IterableIterator<TOutputResult>
}

type CsvProductDescriptor = { title: string }

class ProductCsvImporter extends CsvParser<
  CsvSchema,
  CsvProductDescriptor,
  CsvProductDescriptor
> {
  *readData(
    parsedContent: CsvProductDescriptor[]
  ): IterableIterator<CsvProductDescriptor> {
    for (let i = 0; i < parsedContent.length; ++i) {
      const outputTuple = {} as CsvProductDescriptor

      const tuple = parsedContent[i]
      const tupleKeys = Object.keys(tuple)

      let actualLineNumber = 1
      for (const tupleKey of tupleKeys) {
        const schemaColumn = this.$$schema.columns.find((column) => {
          return column.name === tupleKey
        })
        if (!schemaColumn) {
          throw new Error(
            `Unable to trait column ${tupleKey} from the csv file. No target column found in the provided schema`
          )
        }

        outputTuple[schemaColumn.mapToEntityField] = schemaColumn.validator(
          tuple[tupleKey],
          {
            line: actualLineNumber,
            column: tupleKey,
          },
          this.container
        )
      }

      ++actualLineNumber
      yield outputTuple
    }
  }
}

import { ProductRepository } from "./repositories/product"

export const exampleImportProductFromCsvSchema: CsvSchema = {
  columns: [
    {
      name: "title",
      validator: async (
        title: string,
        context: CsvParserContext,
        container: AwilixContainer
      ): Promise<string> => {
        const productRepository: ProductRepository =
          container.resolve("productRepository")
        const doesProductWithTitleExists = await productRepository.count({
          where: { title },
        })
        if (doesProductWithTitleExists) {
          throw new Error(
            `The title for the product line ${context.line} is already used. Update the value in the column n° ${context.column}`
          )
        }

        /* Some value mutation */
        return (title as string).toLowerCase()
      },
      mapToEntityField: "title",
    },
  ],
}

const csvProductImporter = new ProductCsvImporter(
  {} as AwilixContainer,
  exampleImportProductFromCsvSchema
)

const fakeData = [{ title: "fjzkfjnaefe " }]

const content = fakeData //await csvProductImporter.parse("test")
for (const product of csvProductImporter.readData(content)) {
  console.log(product)
}

```

**Comment by adrien2p** [comment]
Thanks @zakariaelas for the detailed response 😀

**Dependencies**

Indeed the dependsOn is not a required property, it works also as you mentioned if the line is passed as part of the context.
 
But it also depends on the approach, read below and we can decide that late mr 😀

**Global thoughts**

I still think that reading line by line would work, especially if you go with chunks of data. As we know the format of the file, it is pretty easy to find out if we need more chunks to import a line or if we already have enough data to complete the actual line. But some caveats can appears 👇

The only thing that bothered me is the input format. Representing the related objects on the same line can be complex:
-dynamic columns
-very long line
-not human readable which can leads to unforeseen bugs
- ...

Having them on separate line would mean that it is probable that you have to hold some objects until you find the other object on which there belongs. That object could be at the end of the files 😂 I dont think it is a huge problem if we do not endup storing all the object in memory by keeping a reference to them, which could lead to high memory consumption. But generally speaking there is little chance that it happens on all the objects to process.

In that case reading line by line does not make sens anymore, and instead, we should probably parse the file chunk by chunk and re build the object on our hand 🤔 once an object is built we can import it and so on.

The best options would be an import from a json type object, which would solve that problem. In general, the softs are able to export as json as well as csv. It would also avoid us to have to rebuild the object. And just be able to consume it as it comes. The user would have to extract post mapping but it is the same with csv.

@zakariaelas @fPolic wdth?

**Comment by zakariaelas** [comment]
Recaping what we've compiled during our last meeting here:

### Decision

What we propose is to continuously read the csv file using streams (line by line or chunk by chunk) from whatever location (remote or local), parse the file, and finally expose a data structure representing the entire parsed content. In addition, in order to workaround the limitations that we might encounter in the near future, we believe it would make sense to set a limit size for the file so that we minimize the risk of running out of memory.

As things stand right now, we’ve opted to keep the premature optimization to a minimum and just jump right into making something happen. As much as we recognize the shortcomings of the current compromise, we think it’s important to get going and take on any challenges as they arise. For now, we’re going with simplicity by being able to consume the content easily and run aggregations relatively easier instead of optimizing for memory.

### Storing parsed content during `dry_run`

Between the time users initiate a dry run and the time they confirm it, we need to store the parsed content somewhere. We’ve opted on relying on Redis for this one. The parsed content will be stored with a set expiry time (say 1 hour). We believe this is better to say, storing it in a DB, because then we wouldn’t have to schedule another job for cleaning up the content in case a user does not complete the flow and hangs the dry run.

Let me know if we've missed out on anything

cc @olivermrbl @fPolic @adrien2p @pKorsholm

**Comment by adrien2p** [comment]
I would add one point. From what zak told me, tekla has an export that represents approximately 800kb of data. We can limit the files to 10-20mb for example.

To be able to choose the right size, we can benchmark the memory consumption depending on the quantity of data that a file hold. Also, this file size limit could be configurable from the medusa-config file to give more flexibility for the users if they use bigger machine for example.

**Comment by olivermrbl** [comment]
Solved in #1572
